### PR TITLE
https support

### PR DESCRIPTION
**Pull request by srid**
Https is enabled by default. Consequently this adds two new arguments to `ob deploy init`: 

```
  --ssl-hostname SSLHOSTNAME
                           SSL hostname
  --ssl-email SSLEMAIL     SSL admin email
```

- [x] Initial development
- [x] Verify, after subdomain is activated

### CODE REVIEW

**Comment by srid** [comment]
We might want to look into revamping the deployment configuration mechanism. At the moment the deploy directory is growing to hold much more small config files:

```
backend_hosts        config  ssh_key    ssl_host
backend_known_hosts  src     ssl_email
```

**Comment by srid** [comment]
Verified :tada: https://obelisk-skeleton.srid.ca/

**Comment by ryantrinkle** [comment]
@srid Awesome, thanks!  One thing to note: i think we should have a single email address for all alerts, not one specific to SSL.

**Comment by srid** [comment]
@ryantrinkle I've renamed `--ssl-email` to `--admin-email` - so the later can be reused in other places.

File: lib/command/src/Obelisk/Command.hs [hunk 1]
```diff
@@ -161,6 +161,8 @@ deployInitOpts = DeployInitOpts
   <$> strArgument (action "directory" <> metavar "DEPLOYDIR" <> help "Path to a directory that it will create")
   <*> strOption (long "ssh-key" <> action "file" <> metavar "SSHKEY" <> help "Path to an ssh key that it will symlink to")
   <*> some (strOption (long "hostname" <> metavar "HOSTNAME" <> help "hostname of the deployment target"))
+  <*> strOption (long "ssl-hostname" <> metavar "SSLHOSTNAME" <> help "SSL hostname")
+  <*> strOption (long "admin-email" <> metavar "ADMINEMAIL" <> help "Site admin email")
```
Comment (ryantrinkle) [inline lib/command/src/Obelisk/Command.hs]: We need a better description here.  How about "Email address where administrative alerts will be sent"?

File: lib/command/src/Obelisk/Command.hs [hunk 2]
```diff
@@ -161,6 +161,8 @@ deployInitOpts = DeployInitOpts
   <$> strArgument (action "directory" <> metavar "DEPLOYDIR" <> help "Path to a directory that it will create")
   <*> strOption (long "ssh-key" <> action "file" <> metavar "SSHKEY" <> help "Path to an ssh key that it will symlink to")
   <*> some (strOption (long "hostname" <> metavar "HOSTNAME" <> help "hostname of the deployment target"))
+  <*> strOption (long "ssl-hostname" <> metavar "SSLHOSTNAME" <> help "SSL hostname")
```
Comment (ryantrinkle) [inline lib/command/src/Obelisk/Command.hs]: How about "The hostname that end-users will use to access the site"

Also, come to think of it, this should just be taken from the route, shouldn't it?

File: lib/command/src/Obelisk/Command.hs [hunk 3]
```diff
@@ -181,6 +183,8 @@ data DeployInitOpts = DeployInitOpts
   { _deployInitOpts_outputDir :: FilePath
   , _deployInitOpts_sshKey :: FilePath
   , _deployInitOpts_hostname :: [String]
+  , _deployInitOpts_sslHostname :: String
+  , _deployInitOpts_adminEmail :: String
```
Comment (ryantrinkle) [inline lib/command/src/Obelisk/Command.hs]: What is with all the use of `String` here?  Why not `Text`?

Comment (3noch) [reply lib/command/src/Obelisk/Command.hs]: This whole code base could use a revamp to use `Text`. `process` uses `String` for all its arguments and I think that bled all the way up to our types.

Comment (srid) [reply lib/command/src/Obelisk/Command.hs]: Yup, I used `String` only to be consistent. I can open a separate PR that converts the whole code base to use `Text`.

Comment (ryantrinkle) [reply lib/command/src/Obelisk/Command.hs]: @3noch I understand how that can happen, but it's a symptom of using the wrong approach.  People should always keep the core of their code in the best types available, and always do conversions to worse types at the edge of the system.  In other words, convert to "better" types at the earliest possible time, and convert to "worse" types at the last possible moment.  So, `String` should never pollute datastructures like this that we are writing.

@srid That sounds excellent; thanks!

Comment (ryantrinkle) [reply lib/command/src/Obelisk/Command.hs]: @srid And I 100% agree about keeping consistent with the surrounding code.

Comment (3noch) [reply lib/command/src/Obelisk/Command.hs]: @ryantrinkle I very much agree. I was tempted to make this transition multiple times but realized that using both `Text` and `String` in unpredictable places would be worse than just being consistent and doing a big switch at some point.

File: lib/command/src/Obelisk/Command.hs (hunk 2 above)
Comment (ryantrinkle) [reply lib/command/src/Obelisk/Command.hs]: @srid That sounds good.  Let's make sure we cover all our bases, e.g. if someone passes an invalid URL or a URL with a path, etc.

File: default.nix [hunk 4]
```diff
@@ -168,16 +168,16 @@ rec {
       ln -s ${compressedJs frontend} $out/frontend.jsexe
     ''; #TODO: run frontend.jsexe through the asset processing pipeline
 
-  server = exe: hostName:
+  server = exe: hostName: sslHost: adminEmail:
```
Comment (ryantrinkle) [inline default.nix]: I don't think we should just add more params here.  Let's make it use a record or something instead.

File: default.nix [hunk 5]
```diff
@@ -168,16 +168,16 @@ rec {
...
-          # sslConfig = {
-          #   hostName = "example.com";
-          #   adminEmail = "webmaster@example.com";
-          #   subdomains = [ "www" ];
-          # };
+          sslConfig = {
+            hostName = sslHost;
+            adminEmail = adminEmail;
+            subdomains = [ ];
+          };
```
Comment (ryantrinkle) [inline default.nix]: It looks like this makes HTTPS *mandatory*.  It should not be mandatory - HTTPS termination doesn't make sense, e.g., if the server is running behind a load balancer.

Comment (srid) [reply default.nix]: Pushed a change that includes `--disable-https` to disable the by-default https behaviour.

Comment (ryantrinkle) [reply default.nix]: We should not use "" to denote a missing sslHost; it should be `null`.

File: default.nix [hunk 6]
```diff
@@ -168,17 +168,22 @@ rec {
...
+        # https is enabled unless sslHost is empty string
+        extraConfig = if sslHost == "" then {} else {
+          sslConfig = {
+            hostName = sslHost;
+            adminEmail = adminEmail;
+            subdomains = [ ];
+          };
         };
+        httpsConfig = {
+          backendPort = 8000; # TODO: read from config
```
Comment (ryantrinkle) [inline default.nix]: This todo looks wrong.

File: default.nix [hunk 7]
```diff
@@ -168,17 +168,22 @@ rec {
...
+        } // extraConfig;
+        # TODO: switch from lib/https to nixos builtin ssl config
```
Comment (ryantrinkle) [inline default.nix]: Please make a ticket for this one.

File: lib/command/src/Obelisk/Command.hs [hunk 8]
```diff
@@ -161,6 +161,9 @@ deployInitOpts = DeployInitOpts
   <$> strArgument (action "directory" <> metavar "DEPLOYDIR" <> help "Path to a directory that it will create")
   <*> strOption (long "ssh-key" <> action "file" <> metavar "SSHKEY" <> help "Path to an ssh key that it will symlink to")
   <*> some (strOption (long "hostname" <> metavar "HOSTNAME" <> help "hostname of the deployment target"))
+  <*> strOption (long "route" <> metavar "PUBLICROUTE" <> help "Public URL to your app (comon/route)")
```
Comment (ryantrinkle) [inline lib/command/src/Obelisk/Command.hs]: Let's take out `(common/route)` here.  Without further explanation, that won't mean anything to people.

File: lib/command/src/Obelisk/Command.hs [hunk 9]
```diff
@@ -181,6 +184,9 @@ data DeployInitOpts = DeployInitOpts
   { _deployInitOpts_outputDir :: FilePath
   , _deployInitOpts_sshKey :: FilePath
   , _deployInitOpts_hostname :: [String]
+  , _deployInitOpts_route :: String
+  , _deployInitOpts_adminEmail :: String
+  , _deployInitOpts_disableHttps :: Bool
```
Comment (ryantrinkle) [inline lib/command/src/Obelisk/Command.hs]: This should be `enableHttps` here

File: lib/command/src/Obelisk/Command/Deploy.hs [hunk 10]
```diff
@@ -116,6 +140,9 @@ deployPush deployPath getNixBuilders = do
           ["-C", deployPath, "add", "."]
         callProcessAndLogOutput (Debug, Error) $ proc "git"
           ["-C", deployPath, "commit", "-m", "New deployment"]
+    putLog Notice $
+      (if disableHttps then "Deployed (without https) => " else "Deployed => ")
```
Comment (ryantrinkle) [inline lib/command/src/Obelisk/Command/Deploy.hs]: I don't think we need to mention https here.

File: lib/command/src/Obelisk/Command.hs (hunk 2 above)
Comment (srid) [reply lib/command/src/Obelisk/Command.hs]: Here's the full range of possible route validation:

```haskell
    InvalidRoute_MissingScheme uri -> route uri "must have an URI scheme"
    InvalidRoute_NotHttps uri -> route uri "must be HTTPS"
    InvalidRoute_MissingHost uri -> route uri "must contain a hostname"
    InvalidRoute_HasPort uri -> route uri "cannot specify port"
    InvalidRoute_HasPath uri -> route uri "cannot contain path"
 ```

File: default.nix (hunk 5 above)
Comment (srid) [reply default.nix]: Done.

**Comment by ryantrinkle** [comment]
@srid How can I upgrade an existing obelisk project to use this?

**Comment by srid** [comment]
> @srid How can I upgrade an existing obelisk project to use this?

You might want to test/use #77 instead of this PR.

In any case, try this from the deploy directory of that project:

```
echo True > enable_https
echo admin@reflex-frp.org > admin_email
echo https://reflex-frp.org/ > config/common/route
git add ... && git commit ...
ob deploy push
```

**Comment by srid** [comment]
Superseded by #97

**Pull request closed without merging**